### DISCUSSION TITLE
Randomly stopping with "ALARM" but no reason

### DISCUSSION THREAD

**Discussion by marshr32** [winder/Universal-G-Code-Sender#2657] [Help]
Hi, 

I'm having an issue where my machine will randomly stop halfway through a job, requiring a controller reset and a restart of the job.
There is no limit switch being hit or anything like that.
It seems to be getting more frequent.

I've attached a log file, can anyone see anything to tell me what's going wrong?

Is it losing comms or something?
I've changed USB cables and taken a USB hub out of the chain, but it still happens.

Thanks.
[messages1.log](https://github.com/user-attachments/files/18086793/messages1.log)

**Comment by breiler** [comment]
This sounds like a EMI problem, what type of controller do you have?

**Reply by marshr32** [reply]
It's an Arduino with GRBL.

Thing is, I've been running this machine for years and not had this many problems.

It also seems to be more frequent with larger files (200,000 lines) stopping every 2/3000 lines. 
Could it be a memory problem with UGS or something like that?

Not blaming UGS, just trying to find solutions :)

**Reply by breiler** [reply]
How have you connected things to the Arduino, are you using a shield?
What Alarm do you get?

I am ruling out UGS in this case since it never sends an Alarm. UGS receives it from the controller, pauses the file stream and displays it.

As far as you've described makes me think that it is a problem with EMI which triggers the limit switch inputs. It can either occur immediately when a high(er) power equipment is turned on such as spindles, shopvacs, pumps. Or it can build up such as with static electricity with vacuum hoses.

It depends a bit on what Alarm you are getting from the controller. Some Ardunios, especially the cheap Chinese clones  using the CH340 serial drivers, are mentioned in numerous error reports about being unstable and may corrupt the data being sent through the USB data line. This will cause random alarms about the Gcode being faulty.

I have a test machine in my office where I can connect a controller based on this board. It is very cheap which unfortunately attracts new users into buying it. But it has no filtering or isolating circuitry for the inputs, which (in my opinion) makes this board useless for CNC machines. 
![image](https://github.com/user-attachments/assets/96ac97f4-09ee-44de-9950-bf031af0691c)

**Comment by marshr32** [comment]
Hi Breiler, thanks for your help and sorry for the slow responses...... it's a busy time leading up to christmas.

The Arduino is a genuine one, using a screw terminal shield to connect shielded cables with suppression caps across them for the limit switches, but not other inputs like the out of place alarm for my closed loop steppers and my pause button. Maybe I should look at that.
Also thinking I should shield the plastic box the controller is in.

I have always had a problem where I can't use the touchpad on my laptop once the spindle starts, due to the mouse jumping all over the show when I touch the touchpad. A USB mouse solves that though. 
So I guess EMI is at least part of the issue.

The alarm, is just "alarm" with no reason given, as opposed to a limit switch or the door (pause) button being hit where it shows up in UGS.
Once I do the soft reset, it's gone and I'm none the wiser.
This is why it's frustrating, and I was hoping the log would show something.

Good to know it's probably not UGS, though I think it's interesting that the stops seem to happen more frequently with longer Gcode files. 
And that it's becoming more frequent.

If you have any other ideas on how I might help it, then please let me know.
Otherwise, I'll have a look at EMI issues in the new year and report back.

Merry Christmas :)

**Reply by breiler** [reply]
> I have always had a problem where I can't use the touchpad on my laptop once the spindle starts, due to the mouse jumping all over the show when I touch the touchpad.

The touch pads in laptops nowdays are usually capacitive which are more susceptible to EMI. I have some issues with this my self, but it got a lot better when I installed a properly shielded cable for my spindle.

I think that this post is a good read on how to deal with shielding and grounding:
https://www.cnccookbook.com/cnc-electrical-noise-grounding-techniques-interference-filter/

**Comment by marshr32** [comment]
So hopefully I have figured this out....... further testing to confirm.

It looks like a connector on one of my limit switches was loose.
I use NC wiring for the switches, so a quick interruption pauses the controller but doesn't show the switch as being triggered after the soft reset, like it normally would when the switch is actually hit.
Is triggering of limit switches recorded in the logs at all? 
I couldn't find it, and it'd be a useful troubleshooting tool.
Or is there an option of what to log?

When a limit switch is triggered the alarm message comes up straight away, but the details of which switch it is doesn't immediately show. I have to do a soft reset to see which switch is triggered, is that normal operation?
